pkg/storage/tsdb/ref_cache.go

```diff
@@ -131,7 +129,7 @@ func (s *refCacheStripe) purge(keepUntil time.Time) {
 		// Since we do expect very few fingerprint collisions, we
 		// have an optimized implementation for the common case.
 		if len(entries) == 1 {
-			if entries[0].touchedAt.Before(keepUntil) {
+			if entries[0].touchedAt < keepUntil.UnixNano() {
```

Comment:
There's a little bit of math involved each time we call `.UnixNano()`. Since we do it once for each entry and we know we have 1M+ entries per ingester, we could store the result of `.UnixNano()` at the beginning of the function.